### PR TITLE
Interfaces Made More Generic

### PR DESCRIPTION
**Pull request by wchen329**
Polygon, Line and Point now no longer CFPoint, CFLine, and CFPolygon which is more in line with it being an interface. Tests have been modified to use the interfaces as well.

In addition, the Point is no longer immutable.

### CODE REVIEW

**Review by skaymen — APPROVED** [review]
These changes all seem very reasonable to me. Nice work on this!

**Comment by wchen329** [comment]
Thanks Stan! These should be more useful to other conventions now.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPolygon.java [hunk 1]
```diff
@@ -96,11 +96,23 @@ public void setData(Array data) {
 		this.data = data;
 	}
 	
+	/**
+	 * Sets the next polygon which make up the multipolygon which this polygon is a part of.
+	 * If next is a CFPolygon, automatically connects the other polygon to this polygon as well.
+	 */
+	public void setNext(Polygon next) {
+		if(next instanceof CFPolygon) {
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPolygon.java:104]: good use of `instanceof`

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java [hunk 2]
```diff
@@ -27,9 +27,15 @@ public Polygon readPolygon(String name, int index) {
...
 		if(polyvar == null) return null;
 		
-		// create a blank cf Polygon
-		Polygon poly = new CFPolygon();
-		return poly.setupPolygon(ds, polyvar, index);
+		Polygon poly = null;
+		
+		// CFConvention
+		if(ds.findGlobalAttribute(CF.CONVENTIONS) != null)
+			if(ucar.nc2.dataset.conv.CF1Convention.getVersion(ds.findGlobalAttribute(CF.CONVENTIONS).getStringValue()) >= 0)
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java]: Based on my read, `CF1Convention.getVersion()` is delegating to `CF1Convention.extractVersion()`, which looks like it's returning the minor version of the CF convention. By "minor version" I mean the version after the decimal, as opposed to the major version which comes before the decimal.

Format:
MAJOR.MINOR

Example:
1.0

Parsed Example:
MAJOR = 1
MINOR = 0

CF support for Simple Geometries begins to be supported in CF 1.8.
Is zero the right value to compare against to condition the creation of CF 1.8-specific functionality?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java [hunk 3]
```diff
@@ -27,9 +27,15 @@ public Polygon readPolygon(String name, int index) {
...
+				poly = new CFPolygon();
+		
+		if(poly == null) return null;
+		else return poly.setupPolygon(ds, polyvar, index);
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java]: What small changes higher up in the method would enable you to replace line 37 and 38 with a simple `return poly;`?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java [hunk 4]
```diff
@@ -61,30 +73,24 @@ public Point readPoint(String name, int index) {
...
 	
 	/**
 	 * Constructs a new Simple Geometry Reader over the specified dataset.
-	 * Assumes the CF Convention.
 	 * 
 	 * @param ds - the specified dataset
 	 */
 	public SimpleGeometryReader(NetcdfDataset ds) {
 		this.ds = ds;
 	}
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java:95]: Good call on eliminating this constructor. I think it makes sense to assume CF.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java [hunk 5]
```diff
@@ -222,4 +249,16 @@ public CFPoint(double x, double y, CFPoint prev, CFPoint next, Array data) {
 		this.y = y;
 		this.data = data;
 	}
+	
+	/**
+	 * Constructs a new empty point at (0,0) with no connections.
+	 * 
+	 */
+	public CFPoint() {
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java:257]: Have you thought about calling the other CFPoint constructor here?

**Review by cschroedl-gov — CHANGES_REQUESTED** [review]
Looking good! This is an excellent transition to using the interface types. A few small things to consider and a couple tweaks are all that lie between you and Mergetown, USA.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java (hunk 4 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java:95]: Thanks, it also allows for being extended to other standards as well!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java (hunk 2 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryReader.java]: Nice catch, originally it was to allow for any CF-1 Convention, but CF-1.8+ are the only ones with simple geometry support. Will be changed to check for 8!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPolygon.java (hunk 1 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPolygon.java:104]: Thanks! One of those rare uses of it, cast safety.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java (hunk 5 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java:257]: Good POINT, could probably get away with constructor composition here.

Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java:257]: 😆

**Review by cschroedl-gov — APPROVED** [review]
Lookin goodly! When Travis says it is happy, then I'd say these changes have earned a passage on the next ship to Mergetown, USA.

**Pull request merged**